p2p/p2p.md

```diff
@@ -0,0 +1,224 @@
+
+The new peers are tracked by subscribing to `event.EvtPeerConnectednessChanged{}`.
+
+The peer tracker also runs garbage collector (gc) that removes the disconnected peers (determined as disconnected for more than `maxAwaitingTime` or connected peers whose scores are less than or equal to `defaultScore`) from the tracked peers list once every `gcPeriod`.
+
+The peer tracker also provides a block peer functionality which is used to block peers that send invalid network headers. Invalid header is a header that fails when `Verify` method of the header interface is invoked.
+
+The exchange client implements the following interface:
+
+```
+// Getter contains the behavior necessary for a component to retrieve
+// headers that have been processed during header sync.
+type Getter[H Header[H]] interface {
+ Head[H]
+
+ // Get returns the Header corresponding to the given hash.
+ Get(context.Context, Hash) (H, error)
+
+ // GetByHeight returns the Header corresponding to the given block height.
+ GetByHeight(context.Context, uint64) (H, error)
+
+ // GetRangeByHeight returns the given range of Headers.
+ GetRangeByHeight(ctx context.Context, from, amount uint64) ([]H, error)
+
+ // GetVerifiedRange requests the header range from the provided Header and
+ // verifies that the returned headers are adjacent to each other.
+ GetVerifiedRange(ctx context.Context, from H, amount uint64) ([]H, error)
+}
+
+// Head contains the behavior necessary for a component to retrieve
+// the chain head. Note that "chain head" is subjective to the component
+// reporting it.
+type Head[H Header[H]] interface {
+ // Head returns the latest known header.
+ Head(context.Context, ...HeadOption[H]) (H, error)
+}
+```
+
+`Head()` method requests the latest header from trusted peers. The `Head()` requests utilizes 90% of the set deadline (in the form of context deadline) for requests and remaining for determining the best head from gathered responses. The `Head()` call also allows passing an optional `TrustedHead` which allows the caller to specify a trusted header against which the untrusted headers received from a list of tracked peers (limited to `maxUntrustedHeadRequests` of 4) can be verified against, in the absence of trusted peers. Upon receiving headers from peers (either trusted or tracked), the best head is determined as the head:
+
+* with max height among the received
+* which is received from at least `minHeadResponses` of 2 peers
+* when neither or both conditions meet, the head with highest height is used
+
+Apart from requesting the latest header, any arbitrary header(s) can be requested (with 3 retries) using height (`GetByHeight`), hash (`Get`), range (`GetRangeByHeight` and `GetVerifiedRange`) from trusted peers as defined in the request proto message:
+
+```
+message HeaderRequest {
+  oneof data {
+    uint64 origin = 1;
+    bytes hash = 2;
+  }
+  uint64 amount = 3;
+}
+```
```

Comment:
[blocking]
Let's separate the protocol from implementation. I suggest adding `### Protocol` paragraph, which contains all the messages and explains the basic flow. The method description of components can then link to it.